src/libtest.rs

```diff
@@ -0,0 +1,173 @@
+#[serde(tag = "event")]
+#[serde(rename_all = "lowercase")]
+/// Suite event
+pub enum SuiteEvent {
+    /// emitted on the start of a test run, and the start of the doctests
+    Started {
+        /// number of tests in this suite
+        test_count: usize,
+    },
+    /// the suite has finished
+    Ok {
+        /// the number of tests that passed
+        passed: usize,
+        /// the number of tests that failed
+        failed: usize,
+        /// number of tests that were ignored
+        ignored: usize,
+        /// number of benchmarks run
+        measured: usize,
+        /// i think this is based on what you specify in the cargo test argument
+        filtered_out: usize,
+        /// how long the suite took to run
+        exec_time: f32,
+    },
+    /// the suite has at least one failing test
+    Failed {
+        /// the number of tests that passed
+        passed: usize,
+        /// the number of tests that failed
+        failed: usize,
+        /// number of tests that were ignored
+        ignored: usize,
+        /// i think its something to do with benchmarks?
+        measured: usize,
+        /// i think this is based on what you specify in the cargo test argument
+        filtered_out: usize,
+        /// how long the suite took to run
+        exec_time: f32,
+    },
+}
+
+#[derive(Debug, PartialEq, Deserialize, Serialize)]
+#[serde(tag = "event")]
+#[serde(rename_all = "lowercase")]
+/// Test event
+pub enum TestEvent {
+    /// a new test starts
+    Started {
+        /// the name of this test
+        name: String,
+    },
+    /// the test has finished
+    Ok {
+        /// which one
+        name: String,
+        /// in how long
+        exec_time: f32,
+        /// what did it say?
+        stdout: Option<String>,
+    },
+    /// the test has failed
+    Failed {
+        /// which one
+        name: String,
+        /// in how long
+        exec_time: f32,
+        /// why?
+        stdout: Option<String>,
+        /// it timed out?
+        reason: Option<String>,
+        /// what message
+        message: Option<String>,
+    },
+    /// the test has been ignored
+    Ignored {
+        /// which one
+        name: String,
+    },
+    /// the test has timed out
+    Timeout {
+        /// which one
+        name: String,
+    },
+}
+
+impl TestEvent {
+    /// Get the name of this test
+    pub fn name(&self) -> &str {
+        let (Self::Started { name }
+        | Self::Ok { name, .. }
+        | Self::Ignored { name }
+        | Self::Failed { name, .. }
+        | Self::Timeout { name }) = self;
+        name
+    }
+
+    /// Get the stdout of this test, if available.
+    pub fn stdout(&self) -> Option<&str> {
+        match self {
+            Self::Ok { stdout, .. } | Self::Failed { stdout, .. } => stdout.as_deref(),
+            _ => None,
+        }
+    }
+
+    /// If this test is in progress.
+    pub fn in_progress(&self) -> bool {
+        matches!(self, Self::Started { .. })
+    }
```

Comment:
I don't think this is too useful, as something like `in_progress` sounds like it could apply to intermediate messages, too, but those don't exist.

Reply:
Yeah, I never actually needed this.